File: optimization-greedy-algorithms-problems-plus/src/algorithms/activity-selection.ts
```typescript
/**
 * src/algorithms/activity-selection.ts
 *
 * This file implements the Activity Selection Problem using a greedy approach.
 *
 * Problem Description:
 * Given a set of `n` activities, each with a start time `s_i` and a finish time `f_i`,
 * select the maximum number of non-overlapping activities that can be performed by a single person or machine.
 * Two activities `i` and `j` are non-overlapping if `s_i >= f_j` or `s_j >= f_i`.
 *
 * Example:
 * Activities: [(1, 4), (3, 5), (0, 6), (5, 7), (3, 9), (5, 9), (6, 10), (8, 11), (8, 12), (2, 14), (12, 16)]
 * Optimal selection: [(1, 4), (5, 7), (8, 11), (12, 16)] - 4 activities
 */

/**
 * Represents an activity with a start and finish time.
 */
export interface Activity {
    id: number | string; // Optional identifier for debugging/tracking
    start: number;
    finish: number;
}

/**
 * Solves the Activity Selection Problem using a greedy approach.
 *
 * The greedy strategy is to always pick the activity that finishes earliest among the
 * activities that are compatible with the previously selected activity.
 *
 * Algorithm Steps:
 * 1. Sort the activities by their finish times in ascending order.
 *    If two activities have the same finish time, their relative order doesn't matter for correctness,
 *    but sorting by start time as a secondary key can make the output deterministic.
 * 2. Select the first activity (which has the earliest finish time). Add it to the result set.
 * 3. Iterate through the remaining sorted activities. For each activity:
 *    If its start time is greater than or equal to the finish time of the last selected activity,
 *    select this activity and add it to the result set.
 * 4. Continue until all activities have been considered.
 *
 * Time Complexity:
 * - Sorting the activities: O(N log N), where N is the number of activities.
 * - Iterating through sorted activities: O(N).
 * - Total Time Complexity: O(N log N) due to sorting.
 *
 * Space Complexity:
 * - Storing sorted activities: O(N) (if a copy is made, or in-place sort requires O(log N) or O(N) depending on sort algorithm).
 * - Storing the result set: O(N) in the worst case (all activities are selected).
 * - Total Space Complexity: O(N).
 *
 * @param activities An array of Activity objects.
 * @returns An array of selected Activity objects that represents the maximum set of non-overlapping activities.
 */
export function selectActivities(activities: Activity[]): Activity[] {
    // Edge case: If there are no activities, return an empty array.
    if (!activities || activities.length === 0) {
        return [];
    }

    // 1. Sort activities by their finish times in ascending order.
    // If finish times are equal, sorting by start time can provide a deterministic order
    // but is not strictly necessary for correctness of maximum activities.
    // Example sort: (a.finish - b.finish) ensures ascending finish times.
    // If finish times are equal, (a.start - b.start) ensures ascending start times.
    const sortedActivities = [...activities].sort((a, b) => {
        if (a.finish !== b.finish) {
            return a.finish - b.finish;
        }
        return a.start - b.start; // Secondary sort key
    });

    // Initialize the result array with the first activity.
    // The first activity in the sorted list always has the earliest finish time,
    // so it's always a valid choice for the first selected activity.
    const selected: Activity[] = [];
    selected.push(sortedActivities[0]);

    // Keep track of the finish time of the last selected activity.
    let lastFinishTime = sortedActivities[0].finish;

    // 3. Iterate through the remaining activities, starting from the second one.
    for (let i = 1; i < sortedActivities.length; i++) {
        const currentActivity = sortedActivities[i];

        // If the current activity's start time is greater than or equal to the
        // finish time of the last selected activity, it means they are non-overlapping.
        // We can select this activity.
        if (currentActivity.start >= lastFinishTime) {
            selected.push(currentActivity);
            // Update the finish time of the last selected activity.
            lastFinishTime = currentActivity.finish;
        }
    }

    // 4. Return the set of selected activities.
    return selected;
}

// --- Brute Force vs. Optimized (Greedy) Discussion ---
/*
* Brute Force Approach:
*   A brute force approach would involve checking all possible subsets of activities
*   to find the largest subset where no activities overlap.
*   This can be done using recursion with backtracking.
*   For N activities, there are 2^N possible subsets. For each subset, we would need to
*   check for overlaps, which takes O(N log N) (after sorting) or O(N^2) (naively).
*   Thus, the brute force complexity would be O(2^N * N log N) or O(2^N * N^2), which is
*   highly inefficient for even moderately sized N.

* Optimized (Greedy) Approach:
*   The greedy approach implemented above significantly reduces complexity by making
*   locally optimal choices (picking the earliest finishing activity) that lead to a
*   globally optimal solution. The correctness of this greedy strategy can be proven
*   using an exchange argument:
*   Suppose there is an optimal solution `O` that does not include the earliest-finishing
*   activity `a_k`. Let `a_j` be the first activity in `O`. Since `a_k` finishes earliest,
*   `f_k <= f_j`. If `a_k` overlaps with no other activities in `O` (except possibly `a_j`),
*   we can replace `a_j` with `a_k` to form a new optimal solution `O'` of the same size,
*   or even a larger size if `a_k` finishes earlier and allows more subsequent activities.
*   This argument shows that there is always an optimal solution that includes the
*   earliest-finishing activity. This property is crucial for greedy algorithms.
*/

// --- ASCII Art Diagram for Activity Selection Logic ---
/*
Consider activities sorted by finish time:

Activity A: [---S_A---F_A---]
Activity B:        [---S_B---F_B---]
Activity C:                    [---S_C---F_C---]
Activity D:     [---S_D---F_D---]  (overlaps with A, B)
Activity E:              [---S_E---F_E---] (overlaps with B, C)

Sorted by finish time: A, D, B, E, C (example, assuming values)

1. Select A. Last finish = F_A
   [---S_A---F_A---]
   Selected: [A]

2. Consider D: S_D < F_A. (D overlaps A). Skip D.
   [---S_A---F_A---]
             [---S_D---F_D---]

3. Consider B: S_B >= F_A. Select B. Last finish = F_B
   [---S_A---F_A---]
             [---S_B---F_B---]
   Selected: [A, B]

4. Consider E: S_E < F_B. (E overlaps B). Skip E.
                     [---S_E---F_E---]

5. Consider C: S_C >= F_B. Select C. Last finish = F_C
             [---S_B---F_B---]
                       [---S_C---F_C---]
   Selected: [A, B, C]

Final selection: [A, B, C]
*/
```

File: data-structures-greedy-algorithms-complete/src/main/java/com/greedy/problems/ActivitySelection.java
```java
package com.greedy.problems;

import com.greedy.utils.Pair;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * **Problem:** Activity Selection Problem
 *
 * Given a set of n activities, each having a start time (s_i) and a finish time (f_i).
 * We need to select the maximum number of non-overlapping activities that can be performed by a single person.
 * Assume that the activities are already sorted by their finish times. If not, we should sort them first.
 *
 * **Example:**
 * Activities: (1, 4), (3, 5), (0, 6), (5, 7), (3, 8), (5, 9), (6, 10), (8, 11), (8, 12), (2, 13), (12, 14)
 * Sorted by finish time: (1, 4), (3, 5), (0, 6), (5, 7), (3, 8), (5, 9), (6, 10), (8, 11), (8, 12), (2, 13), (12, 14)
 *
 * **Greedy Choice Property:**
 * The greedy choice is to always select the activity that finishes earliest among the
 * remaining compatible activities. This is because selecting an activity that finishes
 * early leaves the maximum amount of time available for other activities.
 *
 * **Optimal Substructure:**
 * If the optimal solution for a set of activities S includes activity 'k', then the
 * remaining activities in the optimal solution are themselves an optimal solution
 * for the set of activities compatible with 'k' (i.e., activities that start after
 * 'k' finishes).
 *
 * **Algorithm Steps:**
 * 1. Sort the activities by their finish times in non-decreasing order.
 * 2. Select the first activity (which has the earliest finish time). Add it to the result set.
 * 3. Iterate through the remaining activities:
 *    If the current activity's start time is greater than or equal to the finish time
 *    of the previously selected activity, select it. Add it to the result set and update
 *    the 'last selected activity's finish time'.
 * 4. Continue until all activities are processed.
 */
public class ActivitySelection {

    /**
     * Represents an activity with a start and finish time.
     */
    public static class Activity {
        int id; // Optional: for identification
        int start;
        int finish;

        public Activity(int id, int start, int finish) {
            this.id = id;
            this.start = start;
            this.finish = finish;
        }

        public int getStart() {
            return start;
        }

        public int getFinish() {
            return finish;
        }

        @Override
        public String toString() {
            return "A" + id + "(" + start + "," + finish + ")";
        }
    }

    /**
     * Solves the Activity Selection Problem using a greedy approach.
     *
     * @param activities A list of Activity objects.
     * @return A list of selected activities that are non-overlapping and maximum in count.
     *
     * **Time Complexity:**
     * - Sorting activities: O(N log N), where N is the number of activities.
     * - Iterating through sorted activities: O(N).
     * - Total: O(N log N) dominated by sorting.
     *
     * **Space Complexity:**
     * - O(N) for storing the sorted activities (if a new list is created) and the result list.
     *   If sorting is in-place and result list is smaller, it could be O(K) where K is selected activities.
     */
    public List<Activity> selectActivitiesGreedy(List<Activity> activities) {
        if (activities == null || activities.isEmpty()) {
            return Collections.emptyList();
        }

        // Step 1: Sort activities by their finish times.
        // If finish times are equal, sort by start times to ensure consistent behavior
        // (though for activity selection, primary sort by finish time is sufficient).
        List<Activity> sortedActivities = new ArrayList<>(activities);
        sortedActivities.sort(Comparator
                .comparingInt(Activity::getFinish)
                .thenComparingInt(Activity::getStart)); // Tie-breaking by start time

        List<Activity> selected = new ArrayList<>();

        // Step 2: Select the first activity (which has the earliest finish time).
        selected.add(sortedActivities.get(0));
        int lastFinishTime = sortedActivities.get(0).getFinish();

        // Step 3 & 4: Iterate through the remaining activities.
        for (int i = 1; i < sortedActivities.size(); i++) {
            Activity currentActivity = sortedActivities.get(i);

            // If the current activity's start time is greater than or equal to
            // the finish time of the previously selected activity, select it.
            if (currentActivity.getStart() >= lastFinishTime) {
                selected.add(currentActivity);
                lastFinishTime = currentActivity.getFinish();
            }
        }

        return selected;
    }

    /**
     * An alternative "brute-force" approach (conceptual, not truly implemented for exponential search).
     * This method outlines a recursive backtracking approach to find all possible non-overlapping
     * subsets and then picking the largest one. This is highly inefficient compared to the greedy approach.
     *
     * **Time Complexity:** O(2^N) due to exploring all subsets.
     * **Space Complexity:** O(N) for recursion stack.
     *
     * @param activities A list of Activity objects.
     * @return A list of selected activities.
     */
    public List<Activity> selectActivitiesBruteForceConceptual(List<Activity> activities) {
        if (activities == null || activities.isEmpty()) {
            return Collections.emptyList();
        }

        // Sort by start time for consistent exploration, or any order.
        // For actual brute force, sorting isn't strictly necessary but helps
        // avoid redundant processing if combined with memoization.
        List<Activity> sortedActivities = new ArrayList<>(activities);
        sortedActivities.sort(Comparator.comparingInt(Activity::getStart));

        List<Activity> bestSelection = new ArrayList<>();
        findMaxActivitiesRecursive(sortedActivities, 0, new ArrayList<>(), bestSelection);
        return bestSelection;
    }

    /**
     * Helper for brute-force conceptual approach.
     * Recursively tries to build a selection of non-overlapping activities.
     *
     * @param allActivities The list of all available activities.
     * @param currentIndex The current index in 'allActivities' to consider.
     * @param currentSelection The activities selected so far in the current recursive path.
     * @param bestSelection The overall best selection found so far (passed by reference).
     */
    private void findMaxActivitiesRecursive(List<Activity> allActivities, int currentIndex,
                                            List<Activity> currentSelection, List<Activity> bestSelection) {
        // Base case: If all activities have been considered.
        if (currentIndex == allActivities.size()) {
            if (currentSelection.size() > bestSelection.size()) {
                // Update best selection if current one is larger
                bestSelection.clear();
                bestSelection.addAll(currentSelection);
            }
            return;
        }

        Activity current = allActivities.get(currentIndex);

        // Option 1: Try to include the current activity
        boolean canInclude = true;
        if (!currentSelection.isEmpty()) {
            // Check for overlap with the last activity added to currentSelection
            // Note: For this to work correctly, currentSelection would ideally be kept sorted by finish time,
            // or we'd need to check against ALL activities in currentSelection.
            // A simpler non-overlapping check is to just ensure the new activity doesn't conflict with any existing one.
            // This makes the check O(K) where K is size of currentSelection.
            for (Activity selected : currentSelection) {
                if (Math.max(selected.start, current.start) < Math.min(selected.finish, current.finish)) {
                    canInclude = false; // Overlap detected
                    break;
                }
            }
        }

        if (canInclude) {
            currentSelection.add(current);
            findMaxActivitiesRecursive(allActivities, currentIndex + 1, currentSelection, bestSelection);
            currentSelection.remove(currentSelection.size() - 1); // Backtrack
        }

        // Option 2: Exclude the current activity
        findMaxActivitiesRecursive(allActivities, currentIndex + 1, currentSelection, bestSelection);
    }
}

```